hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/MultipartCryptoKeyInputStream.java

```diff
@@ -0,0 +1,373 @@
+    if (lenToRead < numBytesToRead) {
+      // Adjustment required; Update readLengthAdjustedBy.
+      readLengthAdjustedBy = numBytesToRead - lenToRead;
+      LOG.debug("OzoneCryptoInputStream for key: {} part: {} adjusted length " +
+              "by +{} to account for Crypto buffer boundary",
+          key, partIndex, readLengthAdjustedBy);
+    }
+    return numBytesToRead;
+  }
+
+  /**
+   * Seeks the InputStream to the specified position. This involves 2 steps:
+   *    1. Updating the partIndex to the partStream corresponding to the
+   *    seeked position.
+   *    2. Seeking the corresponding partStream to the adjusted position.
+   *
+   * For example, let’s say the part sizes are 200 bytes and part[0] stores
+   * data from indices 0 - 199, part[1] from indices 200 - 399 and so on.
+   * Let’s say we seek to position 240. In the first step, the partIndex
+   * would be updated to 1 as indices 200 - 399 reside in partStream[1]. In
+   * the second step, the partStream[1] would be seeked to position 40 (=
+   * 240 - blockOffset[1] (= 200)).
+   */
+  @Override
+  public void seek(long pos) throws IOException {
+    if (pos == 0 && length == 0) {
+      // It is possible for length and pos to be zero in which case
+      // seek should return instead of throwing exception
+      return;
+    }
+    if (pos < 0 || pos > length) {
+      throw new EOFException("EOF encountered at pos: " + pos);
+    }
+
+    // 1. Update the partIndex
+    if (partIndex >= partStreams.size()) {
+      partIndex = Arrays.binarySearch(partOffsets, pos);
+    } else if (pos < partOffsets[partIndex]) {
+      partIndex =
+          Arrays.binarySearch(partOffsets, 0, partIndex, pos);
+    } else if (pos >= partOffsets[partIndex] + partStreams
+        .get(partIndex).getLength()) {
+      partIndex = Arrays.binarySearch(partOffsets, partIndex + 1,
+          partStreams.size(), pos);
+    }
+    if (partIndex < 0) {
+      // Binary search returns -insertionPoint - 1  if element is not present
+      // in the array. insertionPoint is the point at which element would be
+      // inserted in the sorted array. We need to adjust the blockIndex
+      // accordingly so that partIndex = insertionPoint - 1
+      partIndex = -partIndex - 2;
+    }
+
+    // Reset the previous partStream's position
+    partStreams.get(partIndexOfPrevPos).seek(0);
+
+    // Reset all the partStreams above the partIndex. We do this to reset
+    // any previous reads which might have updated the higher part
+    // streams position.
+    for (int index = partIndex + 1; index < partStreams.size(); index++) {
+      partStreams.get(index).seek(0);
+    }
+    // 2. Seek the partStream to the adjusted position
+    partStreams.get(partIndex).seek(pos - partOffsets[partIndex]);
+    partIndexOfPrevPos = partIndex;
+  }
+
+  @Override
+  public synchronized long getPos() throws IOException {
+    return length == 0 ? 0 : partOffsets[partIndex] +
```

Comment:
Do we need to checkOpen() at the begining?

Reply:
The underlying stream would perform the check. But I will add it here as well.